### DISCUSSION TITLE
Unlock blocks on custom page Checkout impossible

### DISCUSSION THREAD

**Discussion by audi63** [woocommerce/woocommerce#45018] [WooCommerce, Blocks, and Site Editing]
Hi,
I want to move down one of the blocks but unlocking the block checkout seems impossible... children's blocks are locked too..
Can you help me ?

Thanks for your work

**Comment by audi63** [comment]
I want to move the Express checkout under Terms of Service.. I can't find a solution to move it properly... and that's not to mention that I would like to keep Express checkout so that it displays under certain conditions, when certain additional fields are not required or depending on the type of sale.

Does anyone have a solution to offer me?

**Comment by opr** [comment]
Hi there, it's not currently possible to do this from within the editor, however I'm wondering if registering a custom block and making it a child of the terms and conditions block would work for you? You could hide the original block with CSS.

> that's not to mention that I would like to keep Express checkout so that it displays under certain conditions, when certain additional fields are not required or depending on the type of sale.

I understand you want to show/hide the block based on certain conditions? I think that if you take the approach above, you can render your custom block based on any condition you like.

Hope this helps answer your question, but please feel free to reach out again if you need more help!

**Comment by audi63** [comment]
Thank you for your response but I've already tested it, and this solution didn't work.

**Comment by opr** [comment]
Hi @audi63 what exactly did you try and what errors did you experience? If you can share some more context with us about this we can try to debug what the issue is and if we need to make any changes to make it easier for developers to extend the blocks. Thanks!

Reopening until we get more feedback.

**Comment by audi63** [comment]
Thank @opr , All children's blocks are blocked. I can understand that we can't delete some of them but we should be able to organize them as we need, and certain child blocks, like ExpressCheckout and certainly others, we should have the right to move or delete them if we don't want the functionality. I have the same problem if I create a personalized block: the child blocks are always clocked, and no modification is possible.

![block-checkout-custom](https://github.com/woocommerce/woocommerce-blocks/assets/52264873/7c83c183-e0bc-4b29-80a3-f2922f4040ed)
![block-checkout-original](https://github.com/woocommerce/woocommerce-blocks/assets/52264873/89fea327-e5af-4b79-acc8-b62a46c4b367)

**Comment by opr** [comment]
Hi thanks for the additional feedback!

> I have the same problem if I create a personalized block: the child blocks are always clocked, and no modification is possible.

I actually mean if you create a custom block as a child of `woocommerce/checkout` you should be able to move this freely. It is intended and expected that merchants cannot move/remove certain blocks as we want to ensure a consistent user experience for shoppers and merchants when it comes to using the Cart and Checkout blocks.

I'll move this to a discussion as I don't think it's an issue that requires action from us, and I'll also cc @pmcpinto - just to make you aware of this discussion. Thanks!

**Comment by alexflorisca** [comment]
I think we need to find the balance between customisation and ensuring a great UX for customers. If a merchant makes a mistake with customisation for example, and say, removes the address fields, which are necessary to make payment, they may loose a lot of money. So I think it's our responsibility to first and foremost, provide a stable, well converting checkout experience. The customisations needs to fit around this, and naturally there will be some limitations. We will generally aim to allow merchants and developers to add, but not remove things, for the reasons I mentioned above.

Also some of the inner blocks are dynamic. If you don't have any express payment methods configured, the block will not appear. Billing and Shipping Address are also conditionally rendered based on items added to cart and some Woo settings.

**Comment by piezak** [comment]
Hello @alexflorisca, @opr 

Could we explore the possibility of unlocking an option to place an order summary between Customer data section and the ‘Buy Now’ button? Show customer the summary just before confirmation button provide quick overview.
In my case, users are experiencing ‘dead clicks’ while trying to adjust the shipping method at the start of the checkout page. A summary closer to the purchase step could improve usability by reducing these unnecessary clicks.

**Reply by alexflorisca** [reply]
I would be interested in these dead clicks, can you provide a bit more details? The shipping method is halfway down the page and clicking the shipping options will update the totals. So I'm curious where your users are clicking? How would moving the order summary closer to the purchase step (I assume you mean the "Place Order" button) solve this problem?

**Reply by piezak** [reply]
Hello @alexflorisca,

Below you will find a screenshot from my cart. We do not use WooCommerce blocks, and we do not give the user the option to select the shipping there (see 8809).

![IMG_8809](https://github.com/user-attachments/assets/11a2aa22-27bd-4048-801d-3eef9617930b)

On the checkout page, I have tracked dead clicks (see 8811) marked in yellow (“Wysyłka” = “Shipping”). 
1. This scenario occurs when a user wants to change the shipping method to pickup. When they find out this text has no function, they do chaotic scrolling to look for a delivery selection, and/or go back to the cart to check if they missed something, and after some time they get so far on the checkout, sometimes they leave.
2. Sometimes, the user wants to select the shipping option right away because it is visible at the beginning, and they didn’t have a chance to select option before, so they are producing dead clicks.

![IMG_8811](https://github.com/user-attachments/assets/39718f22-1b51-43f3-94cd-1ef06b77da10)

In WooCommerce blocks, I have an option to place the summary only at the beginning or at the end, even below the “Place Order” button. An option to place it between the "Payment Methods" and "Place Order" section would be an amazing solution. (see 8812).

![IMG_8812](https://github.com/user-attachments/assets/153afede-6de6-4eb9-8c4c-e27938489663)

**Reply by alexflorisca** [reply]
Right yeah, I get you now, that makes sense. I'll talk to my team about it and see if it's something we can allow. Thanks for the detailed explanation

**Reply by thoseamongstus** [reply]
> Right yeah, I get you now, that makes sense. I'll talk to my team about it and see if it's something we can allow. Thanks for the detailed explanation

Bro, stop with the checkout gatekeeping. "Consistent experience" - nonsense. This isn't Shopify. Why are we being stopped from editing, it does not at all fit with the ethos of the abilities to modify Woocommerce. If someone messes up their site, that's their problem. Add a 'reset' cart/checkout button in the woocommerce tools and call it a day. Or lock product payment buttons ONLY. The rest of it we should be able to unlock, and at the very least, INSERT. Being restricted to paragraph and image makes zero sense.

And I don't see how in previous version we can pretty much do anything we want with custom hooks, and CSS and all stays cross site consistent, and basically modify an entire store to do whatever, but all of a sudden you draw the line at being able to move an email form below the address if we want to? Pfft. And we can't even inject things as freely. These experimental 'fill' things are a nightmare to work with.

Right now the block cart is the most unusable, incomplete, tedious and annoying experience on Woocommerce. I have absolutely no idea why it's even been implemented. 

First of all, it's absolutely asinine we can't move things, or delete what we need. Nor can we even add things that are vital. Wanna add a back to shop button under the 'place order' button. NOPE! Wanna move the crosssells to BELOW the main column. NOPE! 

Not only that, but don't get me started on the CSS modifications for it. They're some of the most tedious nested elements in Woocommerce, that even trying to apply a custom font weight to a title was a massive PITA. All the selectors are so annoying end up having to use !important for the majority of it, especially with nested full paths instead that require all their previous ones, and then just end up getting all mixed up anyway. And it actually makes it WORSE if we use our own custom additional classes for them. And then we have just random 'divs' with no name floating around and the only way to target it is div > div. Good luck trying to get the list items in cross sell with a simple background and border radius to actually line up correctly, cus it has hardcoded 5% right border, but when you fix it, it ends up being too wide for mobile, so the edges appear off screen, so then you need another billion lines of CSS for each and every single element, all of which can only be edited for a few things, instead of just being able to easily modify parent for something as simple as even the font size. NOR do even the buttons or ANY of the inputs follow a site wide edit, so I have to (again, excruciating) modify the CSS for all buttons and inputs solely for the cart/checkout page. Don't even get me started on even adding something like a simple border radius, and then not only that, but then you come to find the PAYMENT inputs are once again, completely different to the rest of the cart. So it's then TRIPLE the css code just to give my inputs a 12px border radius.


And then I add things like 'Best Sellers' to the below the main cart section, but NONE of the buttons line up if one product title just happens to be longer than the other, and there's no way to edit it like other grids so they align appropriately. Also, they just appear - horribly. And then, oh, I forgot to add custom border radius to every single each individual button, so my single product buttons don't match the variable 'choose options'. Lemme just go through the entire cart code again just to make it consistent cus simply using .button-selector button or type[submit] won't work cus everything has it's own selector. Even the colors don't even match. So now let's add another 30 lines of CSS just to edit the colors for everything and then again for checkout blocks.

On mobile it should be 2 product slider, but nothing like that. At least in old cart we could easily just hook a theme or custom product slider into the bottom of the cart page - with wishlist icon/quick view included. Now I have to restrict 'Best Sellers' to 2 products, because using anything more than 4 just creates a ugly long list of grid products for mobile use. 

We can't even inject things now cus everythings experimental fill, so what's the point. Plus requires react JS over simple PHP. Cool. So I guess the custom free shipping bar I implemented into ExperimentalOrderShippingPackages is gonna be broken soon.

And then, don't get me started on the fact EVERYTHING is custom to the cart. So nothing like wishlist icons, quick view or any other theme items don't appear on any of the product list items. And why would any theme even want to support it when it's no where near as ready or good as the old classic cart/checkout page. 


Anyway. Just my two cents. Repeated some stuff probably but you get the point.

Edit. Oh joy. I just discovered the coupon CSS i applied to my blocks cart, doesn't cross over to the checkout. Time to add another 50 lines of duplicate CSS and more headaches figuring out the exact selectors. woohoo.

**Reply by opr** [reply]
@thoseamongstus before we discuss any technical points I want to make it clear that discussions must be civilised. Comments like "bro stop with the checkout gatekeeping", calling our work "asinine", and the general hostile tone of your message aren't acceptable here.

If you want to engage in this discussion, it needs to be respectful and constructive. We're open to criticism and feedback - that's why these discussion exist - but not when it's delivered disrespectfully.

I'll address your technical concerns below, but future comments in this tone won't receive a response. We're all working toward the same goal of improving WooCommerce.

> I have absolutely no idea why it's even been implemented.

The WooCommerce Cart and Checkout blocks were implemented to solve many issues. In the shortcode experience, we offered extensibility through actions/filters and were pretty happy to add them for any use-case that came up. Over the years, all these extensibility points made it harder and harder to implement new features and make changes to the shortcode experience without breaking backward compatibility. The blocks allow us to re-approach extensibility and ensure we create APIs that are focused and hide implementation details from the end user, something hooks don't do very well. (side note, hooks _definitely_ have their place and are useful in many situations, so I'm not totally dismissing them as an extensibility method).

> Why are we being stopped from editing

We don't want to stop you from editing, but as mentioned above we need to make sure the methods we implement to allow this are sustainable in the long term and don't back us into a corner when we want to make changes.

> The rest of it we should be able to unlock, and at the very least, INSERT. Being restricted to paragraph and image makes zero sense.

We discussed this recently and I _believe_ we agreed that any block should be able to insert in the Cart/Checkout inner block areas 🎉  no ETA on this yet though, sorry.

> And I don't see how in previous version we can pretty much do anything we want with custom hooks, and CSS and all stays cross site consistent, and basically modify an entire store to do whatever, but all of a sudden you draw the line at being able to move an email form below the address if we want to? 

Quoted from Alex above:

> >  think we need to find the balance between customisation and ensuring a great UX for customers. If a merchant makes a mistake with customisation for example, and say, removes the address fields, which are necessary to make payment, they may loose a lot of money. So I think it's our responsibility to first and foremost, provide a stable, well converting checkout experience. The customisations needs to fit around this, and naturally there will be some limitations.

> Pfft. And we can't even inject things as freely. These experimental 'fill' things are a nightmare to work with.

Agreed, they're not excellent and that's the reason they're still marked as experimental because we know it's not the implementation we want to keep and support forever. We are currently trying to figure out ways to improve the developer experience here and we have some work in progress with the Interactivity API which may help us move forward. Until that work is done and we get chance to reassess the Cart/Checkout block architecture slot/fills will remain the only way to achieve this.

> First of all, it's absolutely asinine we can't move things, or delete what we need. Nor can we even add things that are vital. Wanna add a back to shop button under the 'place order' button. NOPE! Wanna move the crosssells to BELOW the main column. NOPE!

The Cart/Checkout blocks are opinionated when it comes to field/checkout step order, this is backed by lots of research from institutes such as Baymard. We're not saying we know better than any specific merchant/developer, but we have designed for what we believe will drive conversion in _most cases_. With that said, if you present clear and sensible use cases for why specific things should be moved, we can certainly have the discussion. I want to reiterate that we are being very selective about what extensibility we offer to prevent ourselves being locked in to specific patterns or supporting suboptimal solutions which is why we don't just have a blanket "unlock everything" button.

> Not only that, but don't get me started on the CSS modifications for it. They're some of the most tedious nested elements in Woocommerce, that even trying to apply a custom font weight to a title was a massive PITA.

The CSS architecture follows WordPress block editor patterns. If you have specific selector issues, open a focused issue with examples. On the subject of payment buttons, yes you're right these are different and that is on purpose because they are supposed to adhere to specific guidelines we give to developers. Ideally all buttons should render with the same dimensions to ensure a consistent appearance.

> And then I add things like 'Best Sellers' to the below the main cart section, but NONE of the buttons line up if one product title just happens to be longer than the other, and there's no way to edit it like other grids so they align appropriately.

Have you submitted an issue detailing this along with reproduction steps? It sounds like something that could be improved on our end, so having an issue to track it would be useful.

> On mobile it should be 2 product slider, but nothing like that. At least in old cart we could easily just hook a theme or custom product slider into the bottom of the cart page - with wishlist icon/quick view included. Now I have to restrict 'Best Sellers' to 2 products, because using anything more than 4 just creates a ugly long list of grid products for mobile use.

I think this is valid feedback and could be detailed in an issue, perhaps as an enhancement request?

> We can't even inject things now cus everythings experimental fill, so what's the point. Plus requires react JS over simple PHP. Cool. So I guess the custom free shipping bar I implemented into ExperimentalOrderShippingPackages is gonna be broken soon.

I'm not sure what you mean by "we can't even inject things now" but I can say we won't remove the functionality for the experimental fills without warning, you'll have plenty of time to adjust your code when we do that. Keep an eye on the developer blog for dev advisories.

> And then, don't get me started on the fact EVERYTHING is custom to the cart. So nothing like wishlist icons, quick view or any other theme items don't appear on any of the product list items. And why would any theme even want to support it when it's no where near as ready or good as the old classic cart/checkout page.

Again, we need to know about specific use cases before we can implement an extensibility point for them. If you know of specific plugins/themes that are struggling to integrate with WooCommerce blocks, and you need their functionality, I'd definitely recommend putting them in touch with us via GitHub to illustrate their use case.

> Edit. Oh joy. I just discovered the coupon CSS i applied to my blocks cart, doesn't cross over to the checkout. Time to add another 50 lines of duplicate CSS and more headaches figuring out the exact selectors. woohoo.

If styles aren't inheriting between cart and checkout, please open an issue with specific examples. I also want to recommend the [WooCommerce Community Slack](https://woocommerce.com/community-slack/) where there are plenty of other helpful developers who may be able to give you a hand with issues you're facing.

One final note: WooCommerce is an open source project, you are welcome to propose changes via pull request.

**Comment by ant-topps** [comment]
I too find the current checkout blocks quite limited in terms of customisation.  Some elements i'd love to turn off / remove but they keep coming back.

I can't Hide them (via a tick box). which may be nice for mobile optimisations. 

It would be nice to have the ability to auto expand, or collapse block elements. ie coupon field,
